src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,251 @@
+    Zion::Int                             # Ionic charge (Z - valence electrons)
+    lmax::Int                             # Maximal angular momentum in the non-local part
+    
+    rgrid::Vector{T}                      # Radial grid
+    drgrid::Vector{T}                     # Radial grid derivative / integration factor
+    
+    vloc::Vector{T}                       # Local potential on the radial grid
+    vloc_interp                           # Local potential interpolator
+    
+    projs::Vector{Vector{Vector{T}}}      # Kleinman-Bylander β projectors: projs[l+1][i]
+    projs_interp                          # Projector interpolator
+    h::Vector{Matrix{T}}                  # Projector coupling coefficients per AM channel: h[l+1][i1,i2]
+    
+    pswfcs::Vector{Vector{T}}             # Pseudo-atomic wavefunctions
+    pswfc_ang_moms::Vector{Int}           # Angular momenta of the pseudo-atomic wavefunctions
+    pswfc_occs::Vector{T}                 # Occupations of the pseudo-atomic wavefunctions
+
+    rhoatom::Vector{T}                    # Pseudo-atomic charge density
+
+    rvlocpzdr::Vector{T}                  # (r_i V_{loc}(r_i) + Z) dr_i
+    r2projsdr::Vector{Vector{Vector{T}}}  # r_j^2 β_{li}(r_j) dr_i
+    
+    identifier::String                    # String identifying the PSP
+end
+
+"""
+    PspUpf(Zion::Number, lmax::Number, rgrid::Vector, drgrid::Vector, vloc::Vector,
+           vloc_interp, projs::Vector{Vector{Vector}}, projs_interp, h::Vector{Matrix},
+           pswfcs::Vector{Vector}, pswfc_ang_moms::Vector, pswfc_occs::Vector, rhoatom::Vector,
+           rvlocpzdr::Vector, r2projsdr::Vector{Vector{Vector}}; identifier="")
+
+Construct a Unified Pseudopotential Format pseudopotential. Currently only
+norm-conserving potentials are supported.
+"""
+function PspUpf(Zion, lmax, rgrid::Vector{T}, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                pswfcs, pswfc_ang_moms, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr;
+                identifier="") where {T}
+    length(projs) == length(h) || error("Length of projs and h do not agree.")
+    length(r2projsdr) == length(h) || error("Length of rgrid * projs * dr and h do not agree.")
+    length(pswfcs) == length(pswfc_ang_moms) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "angular momenta do not agree.")
+    length(pswfcs) == length(pswfc_occs) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "occupations do not agree.")
+
+    return PspUpf{T}(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                     pswfcs, pswfc_ang_moms, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr, identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+
+    unsupported = []
+    pseudo["header"]["core_correction"] && push!(unsupported, "nonlinear core correction")
```

Comment:
Maybe align the expressions better if possible? Maybe use a loop and a dict to map from the key to the longer form you push to make this code clearer.

Reply:
Using the psp type seemed to break some other forward-diff stuff in the tests. I think the `Gplusk_vectors_cart` will have a dual `eltype` when using forward-diff in the model. Should look at taking a type for the pseudo parser?